pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,155 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+from typing import TYPE_CHECKING, Optional
+
+from astroid import nodes
+
+from pylint import checkers, interfaces
+from pylint.checkers import utils
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+_LIST_MODIFIER_METHODS = {"append", "remove"}
+_SET_MODIFIER_METHODS = {"add", "remove"}
+
+
+class ModifiedIterationChecker(checkers.BaseChecker):
+    """Checks for modified iterators in for loops iterations.
+
+    Currently supports `for` loops for Sets, Dictionaries and Lists.
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+
+    name = "modified_iteration"
+
+    msgs = {
+        "W4701": (
+            "Iterated list '%s' is being modified inside for loop body, consider iterating through a copy of it "
+            "instead.",
+            "modified-iterating-list",
+            "Emitted when items are added or removed to a list being iterated through. "
+            "Doing so can result in unexpected behaviour, that is why it is preferred to use a copy of the list.",
+        ),
+        "E4702": (
+            "Iterated dict '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-dict",
+            "Emitted when items are added or removed to a dict being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+        "E4703": (
+            "Iterated set '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-set",
+            "Emitted when items are added or removed to a set being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            message_raised = self._modified_iterating_check(body_node, iter_obj)
+            if not message_raised:  # needs further check when message not raised
```

Comment:
This is me being stupid, but what case is this covering? It's covered by `coveralls` so clearly it's doing something correctly but I can't immediately see what test case this covers.

Reply:
the `_modified_iterating_check` checks if there is 1 of the 3 message to raise on each node in the `for` loop.
If not: it checks also it's children nodes. this covers the "for inside for" test when we want to raise a message for the `append` inside the inner for loop regarding the outer for loop node

Reply:
Ah understood!

What about:
```python
for l in item_list:
     for s in item_set:
         for x in iterator:
             print()
             item_list.append(1)  # [modified-iterating-list]
```
Do we catch that as well? It seems like we can only go down one level right now?

Reply:
I think so too but since it is a false-nagative and more of a rare case I suggest to merge this MR and open a new issue for it.

Reply:
Yes, but the current structure of the code doesn't allow for this. Therefore, I think we should change it now. The following diff fixes the issue:
```diff
diff --git a/pylint/checkers/modified_iterating_checker.py b/pylint/checkers/modified_iterating_checker.py
index 55aeaf518..d745a9606 100644
--- a/pylint/checkers/modified_iterating_checker.py
+++ b/pylint/checkers/modified_iterating_checker.py
@@ -57,10 +57,15 @@ class ModifiedIterationChecker(checkers.BaseChecker):
     def visit_for(self, node: nodes.For) -> None:
         iter_obj = node.iter
         for body_node in node.body:
-            message_raised = self._modified_iterating_check(body_node, iter_obj)
-            if not message_raised:  # needs further check when message not raised
-                for child in body_node.get_children():
-                    self._modified_iterating_check(child, iter_obj)
+            self._check_node_and_children_for_changes_to_iterator(body_node, iter_obj)
+
+    def _check_node_and_children_for_changes_to_iterator(
+        self, body_node: nodes.NodeNG, iter_obj: nodes.NodeNG
+    ) -> None:
+        """See if node or any of its children interact with iterator node."""
+        self._modified_iterating_check(body_node, iter_obj)
+        for child in body_node.get_children():
+            self._check_node_and_children_for_changes_to_iterator(child, iter_obj)
 
     def _modified_iterating_check(
         self, node: nodes.NodeNG, iter_obj: nodes.NodeNG
```

With that change the following test also passes:
```python
# Check for nested for loops and changes to iterators
for l in item_list:
    item_list.append(1)  # [modified-iterating-list]
    for _ in []:
        item_list.append(1)  # [modified-iterating-list]
        for _ in []:
            item_list.append(1)  # [modified-iterating-list]
            for _ in []:
                item_list.append(1)  # [modified-iterating-list]
```

Reply:
nice. verified it works and added this test